spec/std/regex_spec.cr

```diff
@@ -124,6 +124,31 @@ describe "Regex" do
     end
   end
 
+  describe "#match!" do
+    it "returns match data" do
+      md = /(?<bar>.)(?<foo>.)/.match!("Crystal")
+      md[0].should eq "Cr"
+      md.captures.should eq [] of String
+      md.named_captures.should eq({"bar" => "C", "foo" => "r"})
+    end
+
+    it "assigns captures" do
+      md = /foo/.match!("foo")
+      $~.should eq md
+    end
+
+    it "raises on non-match" do
+      expect_raises(Regex::Error) { /Crystal/.match!("foo") }
```

Comment:
It's a good practice to assert the error message as well.

```suggestion
      expect_raises(Regex::Error, "Match not found") { /Crystal/.match!("foo") }
```

Reply:
Actually it's used in two more places:

- https://github.com/crystal-lang/crystal/blob/852a44d5f8fea470acd051ca726940bad3d91f5c/src/regex/pcre.cr#L187
- https://github.com/crystal-lang/crystal/blob/852a44d5f8fea470acd051ca726940bad3d91f5c/src/regex/pcre2.cr#L275

Reply:
File locality doesn't matter. The locations where these other exception instances are raised lie in the call path of `Regex#match!`, so it could be entirely possible that the spec catches either of those.
And it's good practice to *always* validate the type and message of an expected exception.

Reply:
I see. Shouldn't this be something the spec tool enforces, or at least warns about? Because this is my first time hearing about such practice, and I don't believe it's documented either.

Reply:
This is hard to enforce. Some exception types have only a single error message, then it's not necessary to verify.